exp/query/lsm_fetcher.go

```diff
@@ -0,0 +1,177 @@
+	BasePath() string
+}
+
+// NewLSMFetcher creates LSMFetcher without cache.
+// Every call is forwarded to given `upstream` to fetch the LSM data for the tenant.
+func NewLSMFetcher(basePath string, upstream LSMDownloader, log logrus.FieldLogger) *LSMFetcher {
+	return NewLSMFetcherWithCache(basePath, upstream, nil, log)
+}
+
+// NewLSMFetcherWithCache creates LSMFetcher with cache.
+func NewLSMFetcherWithCache(basePath string, upstream LSMDownloader, cache LSMCache, log logrus.FieldLogger) *LSMFetcher {
+	return &LSMFetcher{
+		basePath: basePath,
+		cache:    cache,
+		log:      log,
+		upstream: upstream,
+	}
+}
+
+// Fetch try to fetch LSM data for the given (collection,tenant).
+// It works in following modes
+// 1. If cache is nil, fetch from upstream.
+// 2. If cache is enabled and not found in cache, fetch from upstream.
+// 3. If cache is enabled and found in the cache but local version is outdated, fetch from the upstream and remove old version.
+// 4. if cache is enabled, found in the cache and local version is up to date, return from the cache.
+func (l *LSMFetcher) Fetch(ctx context.Context, collection, tenant string, version uint64) (*lsmkv.Store, string, error) {
+	var (
+		basePath          = l.basePath
+		tenantPath string = ""
+		download   bool   = true
+		oldVersion uint64
+	)
+
+	if l.cache != nil {
+		basePath = l.cache.BasePath()
+		c, err := l.cache.Tenant(collection, tenant)
+		if err == nil && version != 0 && c.Version >= version {
+			tenantPath = c.AbsolutePath()
+			download = false
+		}
+		if err == nil && version != 0 && c.Version < version {
+			// oldVersion is used to cleanup previous version of tenant data.
+			oldVersion = c.Version
+		}
+		if err != nil && !errors.Is(err, ErrTenantNotFound) {
+			return nil, "", err
+		}
+	}
+
+	if download {
+		var err error
+		// TODO(kavi): Make this upstream download call efficient. Currently when say `n` tenants access `Fetch` at the same time causing cache-miss,
+		// then all `n` tenants can trigger upstream download.
+		// Use something like Go's `singleflight`(https://pkg.go.dev/golang.org/x/sync/singleflight) to do one upstream call and share the cache with other in-flight requests.
+		tenantPath, err = l.download(ctx, basePath, collection, tenant, version)
+		if err != nil {
+			return nil, "", err
+		}
+
+		// populate the cache before returning
+		if l.cache != nil {
+			if err := l.cache.AddTenant(collection, tenant, version); err != nil {
+				return nil, "", fmt.Errorf("failed populate the cache: %w", err)
+			}
+		}
+
+		// remove old version
+		if oldVersion > 0 {
+			// NOTE: Removing old version can be tricky, even though all new lookup returns new version path.
+			// There can be some clients still using old versions. We can leave without removing and delete only if eviction is needed.
+			// But still that doesn't remove above stated problem.
+			if l.remove(ctx, basePath, collection, tenant, oldVersion); err != nil {
```

Comment:
Question: could it be feasible to keep track of current search requests and which tenant they're using ? Think an `async.Map` where key = tenant+version and value = atomic of search using it.
 
 On marking for eviction we can forbid new search to that version (or unmark for deletion if the version is still the latest one). Then we trigger eviction on that count hitting 0 and marked for eviction ? With a proper reaper ensuring we don't have runway process holding an eviction happening and we eat the rare failing search. WDYT ? (this can all come as follow up prs, just starting the discussion).

Reply:
Good thinking. One problem with that approach is, we are dealing with "files" not in-memory key-values. So once we return to clients, they can do anything with the local path of that tenant. We have very less control. Basically this part

> keep track of current search requests and which tenant they're using 

is hard in our disk based cache. If we decide to go that path, then we need to track the list of "file handles" for all the files on LSM directory. Worse in my opinion.

> On marking for eviction we can forbid new search to that version

We already do that. Sort of, once the new version of tenant is added, every search after that always returns only the latest copy to any clients (note the `key` in `tenants` map is still `/collection/tenant`). But we return the "local filesystem path" to clients. Which they can store and use it as long as they want. We cannot have any control.

If we want more control, they we have to design cache API that returns contents of each file instead of file path. Would be complex at this point in time I guess.